docs/rfcs/0011-customizable-store-dispatcher.md

```diff
@@ -0,0 +1,71 @@
+---
+layout: page
+title: Allow customisation of the Store dispatcher
+permalink: /rfc/0011-customizable-store-dispatcher
+---
+
+* Start date: 2024-01-29
+* RFC PR: [#5353](https://github.com/mozilla-mobile/firefox-android/pull/5353)
+
+## Summary
+
+In most applications of the `Store`, it is preferable to have reducers perform work on the main thread. This helps to simplify the levels of async tasks that need to be synchronised with other async tasks that are already on the main thread.
```

Comment:
> Is there an example of how staying on a single thread helps synchronize these tasks vs synchronizing them across threads?

I mentioned this in the 'Motivation' section below, is there something I can do to make this more clear?

"Android embedders use the main thread for UI, user-facing, or touch handling work. For example, notifying UI components when IO from storage layers have completed, an engine's tasks that can happen on a separate thread, or global-level state updates for different components to observe."

> Is the goal to get results directly from dispatch, or will observation still be the primary method of reading state?

You will still get the results from observations. Maybe I need to rephrase this paragraph better.. The actions will all happen on the same thread async, but when working with other systems that also have separate threads, there are then more than one-level of async work that needs to be waited on.

An example of this that @titooan  and I have run into recently, is where GeckoView has it's own Gecko Main thread for performing async work that also touches into the Gecko platform layer. We want to wait for the `GeckoResult` to come back from there and have those results along with other results from the Android OS touch events update a Store. So this means that we now have at least three threads that provide results we need to sync before we can perform an action: Gecko Main, Store, Android UI (main).

Since the Android OS touch events happen on the main thread already, having a store that is also running on the same thread reduces that complexity.

Similarly, any of the fragment stores that we have also benefit from this simplification as well.

For historical context, the Store was meant to always run on the main thread, but the separate Store thread was possibly a premature optimization for the BrowserStore, which is fairly valuable today, but using the Store for other parts of the app was not something we considered then. In which case, running it on the main thread would be sufficient.

(this context is good for comments, but not useful for an RFC)

Reply:
I will admit to not being very familiar with how `GeckoResult`s are handled, so would love to learn more about this specific use-case. Was there a specific issue with ordering in the example above that would've been solved by having the Store running on the UI thread?

> Since the Android OS touch events happen on the main thread already, having a store that is also running on the same thread reduces that complexity.

So in terms of simplicity, do we just mean thread switches? It seems like the complexity here should be hidden by coroutine primitives and Store abstractions anyway. That is, I would expect a `GeckoResult` handler to `dispatch` to a Store, and the eventual result of that `dispatch` to be consumed through observation by the View. None of these individual pieces should need to know what thread the others are executing on.

Regardless, I do think performance wins from avoiding context switches and simpler test setups are good things to have. Just trying to more fully understand the motivation 😄

Reply:
> So in terms of simplicity, do we just mean thread switches? It seems like the complexity here should be hidden by coroutine primitives and Store abstractions anyway. That is, I would expect a `GeckoResult` handler to `dispatch` to a Store, and the eventual result of that `dispatch` to be consumed through observation by the View. None of these individual pieces should need to know what thread the others are executing on.
> 
> Regardless, I do think performance wins from avoiding context switches and simpler test setups are good things to have. Just trying to more fully understand the motivation 😄

It's a bit more than that too: in the example above, it would make it easier for us to know that when I invoke two methods both of those will be scheduled in the same order too:

```kotlin
geckoView.asyncApi()
store.dispatch(UpdateAction)
```

Reply:
> Was there a specific issue with ordering in the example above that would've been solved by having the Store running on the UI thread?

There isn't any issue today, but @titooan and I will plan to write a design doc about how we would want to build on top of this work that will make APZ features (bottom toolbar and pull-to-refresh) easier to reason about by having a Store to collect data from various systems so that we can reason about them more easily. Specifically, a lot of the logic in [`NestedGeckoView.kt`](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/NestedGeckoView.kt) that is hard to understand and debug today, so requires a lot of time to understand before development can begin there.

Reply:
With the design changes in the updated RFC, the expectation of these changes should be more clear now so that reading the state (not only by flow observers) right after dispatching an action will produce the state change immediately.

> You will still get the results from observations. Maybe I need to rephrase this paragraph better.. 

Updated.